prereise/gather/demanddata/transportation_electrification/smart_charging.py

```diff
@@ -0,0 +1,365 @@
+    :param int power: charger power, EVSE kW.
+    :param int location_strategy: where the vehicle can charge-1, 2, 3, 4, or 5; 1-home only, 2-home and work related, 3-anywhere if possibile, 4-home and school only, 5-home and work and school.
+    :param str veh_type: determine which category (LDV or LDT) to produce charging profiles for
+    :param str filepath: the path to the nhts mat file.
+    :param int comm_type: the type of Commute
+    :param int trip_strategy: determine to charge after any trip (1) or only after the last trip (2)
+    :return: (*numpy.ndarray*) -- charging profiles.
+    """
+
+    # load NHTS data from function
+    if veh_type.lower() == "ldv":
+        newdata = data_helper.remove_ldt(data_helper.load_data(census_region, filepath))
+
+    elif veh_type.lower() == "ldt":
+        newdata = data_helper.remove_ldv(data_helper.load_data(census_region, filepath))
+
+    # updates the weekend and weekday values in the nhts data
+    newdata = data_helper.update_if_weekend(newdata)
+
+    new_columns = [
+        "Electricity cost",
+        "Battery discharge",
+        "Battery charge",
+        "Battery size",
+    ]
+    newdata = newdata.reindex(list(newdata.columns) + new_columns, axis=1, fill_value=0)
+
+    input_day = data_helper.get_input_day(data_helper.get_model_year_dti(model_year))
+
+    TRANS_charge = np.zeros(24 * len(input_day))
+    data_day = data_helper.get_data_day(newdata)
+
+    daily_vmt_total = data_helper.total_daily_vmt(
+        census_region, comm_type, location_strategy, input_day
+    )
+
+    kwh = kwhmi * veh_range
+    emfacvmt = 758118400
+
+    LOAD = np.ones(24 * len(input_day))
+
+    nd_len = len(newdata)
+
+    for day_iter in range(len(input_day)):
+
+        if day_iter == len(input_day) - 1:
+            LOAD_adj = [
+                LOAD[i] + TRANS_charge[i]
+                for i in range(day_iter * 24, (day_iter + 1) * 24)
+            ]
+            LOAD_adj += [LOAD[i] + TRANS_charge[i] for i in range(24)]
+        else:
+            LOAD_adj = [
+                LOAD[i] + TRANS_charge[i]
+                for i in range(day_iter * 24, day_iter * 24 + 48)
+            ]
+
+        cost = LOAD_adj
+
+        G2Vload = np.zeros((100, 48))
+        individualG2Vload = np.zeros(nd_len, 48)
+
+        i = 0
+
+        while i < nd_len:
+            # zero individual block, this for each vehicle per day
+            individual = 0
+
+            # trip amount for each vehicle
+            total_trips = newdata.iloc[
+                i, newdata.columns.get_loc("total vehicle trips")
+            ]
+
+            if data_day[i] == input_day[day_iter]:
+
+                # only home based trips
+                if (
+                    newdata.iloc[i, newdata.columns.get_loc("why from")]
+                    * newdata.iloc[i + total_trips, newdata.columns.get_loc("why to")]
+                    == 1
+                ):
```

Comment:
There's a lot that's going on inside this block. It seems that the major functions are:
* preparing input data for the optimization
* running the optimization
* extracting the optimization results.

If we can move some/all of these into their own functions, it should make it easier to understand everything that's going on within each function and within the overall block, and make testing each component easier.

Reply:
working on restructuring